### DISCUSSION TITLE
FNIRSI GC-01: EMI activates keyboard

### DISCUSSION THREAD

**Discussion by alfmck** [Gissio/radpro#51] [Q&A]
After flashed radpro I noticed that when my GC-01 is in measurement state, time by time it enters in MODE state.
I checked the signals on buttons wires with oscilloscope and saw that the short pulses to ground induces from high voltage generator. The biggest of them were on MODE button wire. It looks that the button state scanning algorithm sometimes detects these pulses as the MODE button was pressed. My GC-01 has a H614 tube soldered nearby this button. Maybe there was defect in pin9 of CH32F103R8T6 that after trying physically filter that pulse, pin9 on MC became faulty. So, I'm going to reprogram radpro, making button UP to fulfill MODE button function. Thank Gissio that You supplied distributives.

**Comment by Gissio** [comment]
Oh my, the dreaded EMI.

Can you post a screenshot or photo of what you see on the oscilloscope? Does this also occur with the original firmware? Did you try adding shielding?

IMHO this should be solved at the software level. But for that I would need to know what the spikes look like.

**Reply by alfmck** [reply]
I'm unable to measure the signal on SETTINGS/RIGHT button - it is faulty now. The signal on LEFT, UP, DOWN buttons is 10kHz, about 1V to ground from the power supply level and didn't cause problems.
When I used original firmware the  SETTINGS/RIGHT button acts a bit strange - after short press I got switching between table and graph modes; after longer press I came to SETTINGS mode. Switching between table and graph modes also did UP and DOWN buttons.
You shouldn't change the software - maybe my GC-01 is defected, more information from other radpro users needed.

**Reply by Gissio** [reply]
alfmck: I received a similar report from another user with a WCH GC-01, so this is not a particular problem with your device.

It would be * extremely * helpful if you could send me a photo of what the signal looks like at the right/settings button and/or at the other buttons. This would help me fix the problem in software.

You say that the signal on the right/settings key is faulty now. What signal do you see?

One more request: could you post your answer in your native language as well as in English? I'm having a bit of a hard time understanding you and this might help me understand. Thank you!

**Reply by alfmck** [reply]
The right/settings key in MY device is faulty now - the resistance on button wire is about 30 Om to ground (was more than 3 MOm before). So in order to have right/settings function in my device, I use slightly changed radpro soft, where the button UP fulfills RIGHT/SETTING function. (By the way could You change please the Source code (zip) in 2.0rc2 release - it is still from version 2.0rc1).

The pulses on ALL buttons are the images of high voltage generator - the frequency and duration of pulses strictly depends on the HV profile set by the user. I suspect that the pulses could also depend on the radioactive particles randomly coming through the tube - I will try to watch them on oscilloscope.

I want to point out that very high induction field is nearby the RIGHT/SETTING button ONLY - other buttons work properly.

**Reply by Gissio** [reply]
Got it. If it's an induction issue, I can synchronize the keyboard readout to the pulses, which should fix the issue.

But I would need to know how long the pulses are. I would also need to know what the upward and downward pulses look like. That's why a photo from your oscilloscope would be greatly helpful (even if you measure it on the other keys).

Can't you fix the right/settings key problem by replacing the switch?

**Reply by alfmck** [reply]
I made 2 videos with sound: 3171.mov -factory default; 3168.mov - custom settings (PVM frequency 2.5kHz, PVM duty cycle 12% )
The scale is 0.5V/cm. Tested on UP, DOWN, LEFT buttons. The measurement process of higher intensity.
I can't fix right/settings key - replacing of microcontroller needed..
Each of files are about 20Mbit - The system allowed to sent 10 Mbit only. Could You give me another link?

**Reply by Gissio** [reply]
Excellent! Try https://filebin.net/ or https://www.file.io/ .

**Reply by alfmck** [reply]
The bin [i1uo9h0rqvzg01v6](https://filebin.net/i1uo9h0rqvzg01v6) was created

**Reply by Gissio** [reply]
That's greatly helpful.

Can you tell me what time scale you used? I would need to know how long the pulses are!

**Reply by alfmck** [reply]
15mks/cm for factory default, 7mks/cm for custom setting

**Reply by Gissio** [reply]
alfmck: that was extremely helpful. Thank you so much for all your effort!

**Reply by alfmck** [reply]
One more experiment: The induction field (probe close to button SETTINGS/RIGHT in the first part of video and close to the button UP on the end of the video). The scales are 0.5V/cm, 10mks/cm.
The bin [4ibejqxch2fxo267](https://filebin.net/4ibejqxch2fxo267) was created

**Comment by Gissio** [comment]
I think I've read somewhere about a similar problem.

Can other users report?

**Reply by alfmck** [reply]
I've a bit corrected a soft and solved the problem by making the UP key to work as SETTINGS/RIGHT key.

**Comment by Gissio** [comment]
alfmck: it my understanding that you are a native Russian speaker. Could you check the Russian translation in 2.2beta3?

The strings are available here: https://github.com/Gissio/radpro/blob/main/platform.io/src/strings/ru.h

If you are willing to help us with these, please follow the discussion here: https://github.com/Gissio/radpro/discussions/241

**Reply by alfmck** [reply]
I'm not. My native language isn't in the list. EN is much more convenient for me in technical works. Thank You, Gissio, for This great project!

**Comment by lakisixx** [comment]
Hey, I have a question. I'm currently trying to use Renode for a project where I coded scenes for a dashboard. I already have the code tested on hardware but it would be good to have it also on Renode to test with other subsystems. The project was in written in STM32CubeIDE, and I have the .elf file. Its the STM32f469i discovery kit with the integrated display and I see you made the ltdc work. Any suggestions on how I can do this, currently just the window is opened but nothing happens. Thanks in advance!
